### PR TITLE
Fix inconsistent timestamp when concurrent usage

### PR DESCRIPTION
**Pull request by han0110**
With global variable `bulkNow`, we may encounter inconsistent `CreatedAt` and `UpdatedAt` in different rows at the same `BulkInsert` when concurrent usage (Someone might have just called the defer function and set the `bulkNow` to `time.Time{}` while the other one is calling `ObjectToMap`).

So I remove the global variable and need to change the return value of `ObjectToMap` to get the `.Struct.Name` of the field to check if it is `CreatedAt` or `UpdatedAt`, then set the value to `bulkNow` at last moment.

### CODE REVIEW

**Comment by bombsimon** [comment]
Thanks for this, nice catch!

I know I did this in a weird way because I had something in mind where you would be able to set your own time but also have the same time for all inserts. I guess I didn't really thing this through regarding running the function in several go routines. I'm actually not really sure about the thread safety at all for this package. :)

I'll have a look at this today!

File: gorm_bulk.go [hunk 1]
```diff
@@ -94,15 +87,13 @@ func scopeFromObjects(db *gorm.DB, objects []interface{}, execFunc ExecFunc) (*g
 		placeholders      []string
 		groups            []string
 		scope             = db.NewScope(objects[0])
+		// bulkNow holds a global now state and will be used for each records as
```
Comment (bombsimon) [inline gorm_bulk.go]: Since you can no longer inject your own `NowFunc()` (although still manipulate the `gorm.NowFunc()`) you should drop this commend and the variable and just call `gorm.NowFunc()` where it's used.

File: gorm_bulk.go [hunk 2]
```diff
@@ -139,7 +130,14 @@ func scopeFromObjects(db *gorm.DB, objects []interface{}, execFunc ExecFunc) (*g
 		}
 
 		for _, key := range columnNames {
-			objectScope.AddToVars(row[key])
+			// Column CreatedAt and UpdatedAt with zero value will be set to same time
+			if name := row[key].Struct.Name; name == "CreatedAt" || name == "UpdatedAt" {
+				if t, ok := row[key].Field.Interface().(time.Time); ok && t.IsZero() {
+					_ = row[key].Set(bulkNow)
```
Comment (bombsimon) [inline gorm_bulk.go]:
```suggestion
					_ = row[key].Set(gorm.NowFunc())
```

File: gorm_bulk.go [hunk 3]
```diff
@@ -139,7 +130,14 @@ func scopeFromObjects(db *gorm.DB, objects []interface{}, execFunc ExecFunc) (*g
 		}
 
 		for _, key := range columnNames {
-			objectScope.AddToVars(row[key])
+			// Column CreatedAt and UpdatedAt with zero value will be set to same time
+			if name := row[key].Struct.Name; name == "CreatedAt" || name == "UpdatedAt" {
+				if t, ok := row[key].Field.Interface().(time.Time); ok && t.IsZero() {
```
Comment (bombsimon) [inline gorm_bulk.go]: I might be wrong but I'm pretty sure `CreatedAt` and `UpdatedAt` can be pointers (even though they're not in the `gorm.Model` type). If so, this type casting can fail since the value might be nil and not a zero time. If that's the case we still want to set the default bulk time, don't we?

**Review by bombsimon — CHANGES_REQUESTED** [review]
I think this looks super good, just one wish to remove my old verbose comment which is no longer needed and a question (or notice) regarding pointers.

Again, thanks for your contribution! 🚀

File: gorm_bulk.go (hunk 2 above)
Comment (han0110) [reply gorm_bulk.go]: But this may cause the different `CreatedAt` or  `UpdatedAt` for rows in the same `BulkInsert`. To make them consistent, we should get the value at the beginning of func `scopeFromObjects`. Gorm itself does the same thing for consistent `CreatedAt` and `UpdatedAt`(https://github.com/jinzhu/gorm/blob/master/callback_create.go#L32-L48).

I think we should remove the comment but keep the variable.

File: gorm_bulk.go (hunk 3 above)
Comment (han0110) [reply gorm_bulk.go]: Yeah, thanks for the reminder. I didn't consider well about the type of `CreatedAt` and `UpdatedAt`. Let's use the old way as `row[key].IsBlank` to check whether the value is set or not.

Comment (han0110) [reply gorm_bulk.go]: Additional minor changes for better readability:
1. changed the statement from `if` to `switch` 
2. make `row[key]` as variable `field`

File: gorm_bulk.go (hunk 2 above)
Comment (bombsimon) [reply gorm_bulk.go]: You’re right, I’m stupid. That was the whole idea behind this in the first place. :)

Keep the variable, dump the comment. Maaaybe keep a comment about why it’s there, like *nowFunc is stored first to result in the same time for every record* or whatever. You can decide if and what! :)

**Comment by bombsimon** [comment]
So I wrote a test for this because it seems it was missing and that was a good idea.

```text
- (time.Time) 2020-01-23 07:58:28.522377 +0100 CET m=+0.002462155,
+ (time.Time) 0001-01-01 00:00:00 +0000 UTC,
  (string) (len=11) "this is foo",
- (time.Time) 2020-01-23 07:58:28.522377 +0100 CET m=+0.002462155
+ (time.Time) 0001-01-01 00:00:00 +0000 UTC
```

Since I got the empty value I tried to debug what we're setting and what the field value ended up being:

```go
if field.IsBlank {
    _ = field.Set(bulkNow)

    fmt.Println(bulkNow)
    fmt.Println(field.Field.Interface())
}
```

This ended up giving me:

```text
2020-01-23 07:58:28.52324 +0100 CET m=+0.003324871
0001-01-01 00:00:00 +0000 UTC
```

So I thought, let's error check this:

```go
if err := field.Set(bulkNow); err != nil {
    panic(err)
}
```

And lo and behold, this gives me:

```text
panic: using unaddressable value [recovered]
        panic: using unaddressable value
```

I don't really have the time to dig into this now but would be nice to adress this before merging!

**Comment by han0110** [comment]
Sorry for causing the bug, let me fix it...

**Comment by bombsimon** [comment]
No worries at all, this is a great catch! I've created a new test so you can do som TDD if you want to! We still want these tests when you're all finished! I'll try to help later if you need but right now I need to work. Please add this test!

```diff
diff --git a/gorm_bulk_test.go b/gorm_bulk_test.go
index dedbc65..d4ed629 100644
--- a/gorm_bulk_test.go
+++ b/gorm_bulk_test.go
@@ -430,3 +430,103 @@ func Test_columnOrder(t *testing.T) {
 		})
 	}
 }
+
+func Test_Timestamps(t *testing.T) {
+	db, _, err := sqlmock.New()
+	require.NoError(t, err)
+
+	gdb, err := gorm.Open("mysql", db)
+	require.NoError(t, err)
+
+	pastDate := time.Date(1985, 1, 1, 0, 0, 0, 0, time.UTC)
+	testStartedAt := time.Now()
+
+	cases := []struct {
+		description string
+		slice       []interface{}
+		expectedSQL string
+		validFunc   func(t *testing.T, vars []interface{})
+	}{
+		{
+			description: "set dates will persist",
+			slice: []interface{}{
+				struct {
+					Foo       string
+					CreatedAt time.Time
+					UpdatedAt time.Time
+				}{
+					Foo:       "this is foo",
+					CreatedAt: pastDate,
+					UpdatedAt: pastDate,
+				},
+			},
+			expectedSQL: "INSERT INTO `` (`created_at`, `foo`, `updated_at`) VALUES (?, ?, ?)",
+			validFunc: func(t *testing.T, vars []interface{}) {
+				createdAt, updatedAt := vars[0], vars[2]
+				assert.Equal(t, pastDate, createdAt)
+				assert.Equal(t, pastDate, updatedAt)
+			},
+		},
+		{
+			description: "empty time pointers will be set to now func",
+			slice: []interface{}{
+				struct {
+					Foo       string
+					CreatedAt *time.Time
+					UpdatedAt *time.Time
+				}{
+					Foo: "this is foo",
+				},
+			},
+			expectedSQL: "INSERT INTO `` (`created_at`, `foo`, `updated_at`) VALUES (?, ?, ?)",
+			validFunc: func(t *testing.T, vars []interface{}) {
+				createdAt, updatedAt := vars[0], vars[2]
+
+				createdAtTime, ok := createdAt.(time.Time)
+				assert.True(t, ok)
+				assert.True(t, createdAtTime.After(testStartedAt))
+
+				updatedAtTime, ok := updatedAt.(time.Time)
+				assert.True(t, ok)
+				assert.True(t, updatedAtTime.After(testStartedAt))
+			},
+		},
+		{
+			description: "empty time will be set to now func",
+			slice: []interface{}{
+				struct {
+					Foo       string
+					CreatedAt time.Time
+					UpdatedAt time.Time
+				}{
+					Foo: "this is foo",
+				},
+			},
+			expectedSQL: "INSERT INTO `` (`created_at`, `foo`, `updated_at`) VALUES (?, ?, ?)",
+			validFunc: func(t *testing.T, vars []interface{}) {
+				createdAt, updatedAt := vars[0], vars[2]
+
+				createdAtTime, ok := createdAt.(time.Time)
+				assert.True(t, ok)
+				assert.True(t, createdAtTime.After(testStartedAt))
+
+				updatedAtTime, ok := updatedAt.(time.Time)
+				assert.True(t, ok)
+				assert.True(t, updatedAtTime.After(testStartedAt))
+			},
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.description, func(t *testing.T) {
+			scope, err := scopeFromObjects(gdb, tc.slice, InsertFunc)
+
+			require.NotNil(t, scope)
+			require.NoError(t, err)
+
+			assert.Equal(t, tc.expectedSQL, scope.SQL)
+
+			tc.validFunc(t, scope.SQLVars)
+		})
+	}
+}
```

**Comment by han0110** [comment]
I finally found the difficulty of `BulkInsert`. IMO, the records passed to `BulkInsert` should be slice of addressable struct (slice of pointer of struct), so the records can be updated after records inserted with some default value set in database. Gorm use callback `forceReloadAfterCreateCallback` to achieve this when we call `Create`. However, when it comes to `BulkInsert`, it's really hard to reload those records and set it back. So I give up supporting that and just focusing on **creating the right records**.

What causing the bug is that the records passed to `BulkInsert` are all slice of struct, which is not addressable. So I make some changes:

1. take the value of `field` first and modify only the value when needed to fix the bug
2. add your test cases
3. add test case to test whether `all records with empty time will be set to the same value`

**Comment by bombsimon** [comment]
Thank you so much for the investigation! If the gorm interface assumes pointers then that would be better to support. The `MyTypeToInterfaceSlice` could be responsible for creating pointers for each element added to the slice, then the end user may work with any preferred type but this package ensures that pointers are created.

It will be a bit too much to refactor since all the tests needs to be re-written so I totally get that you don't mind doing it in this PR.

Feel free to create an issue summarising what you've found out and I'll try to fix that as soon as possible!

**Review by bombsimon — APPROVED** [review]
Lovely! Thanks!

**Comment by han0110** [comment]
Thanks for your patience for reviewing!
